app/services/account_builder.rb

```diff
@@ -0,0 +1,185 @@
+
+  # Overridable method that permits subclass objects to whitelist or modify
+  # params as needed. Returns hash of potentially modified params.
+  # Account can potentially be nil if a `new` form hasn't been submitted yet.
+  def parse_params_for_update(account, params, key)
+    params || {}
+  end
+
+  # Returns a valid account type string (classified) or raises an error;
+  # supports both `build` and `update`.
+  def valid_account_type!(account, account_type, facility)
+    if account.present?
+      valid_account_type_for_existing!(account)
+    else
+      valid_account_type_for_new!(account_type, facility)
+    end
+  end
+
+  # Expects `options[:account_type]` and `options[:facility]`; returns
+  # account_type class name string if account type is valid; otherwise raises
+  # an AccountController::RoutingError exception.
+  def valid_account_type_for_new!(account_type, facility)
+    account_type = account_type.to_s.classify
+    valid_account_types = Account.config.account_types_for_facility(facility)
+    return account_type if valid_account_types.include?(account_type)
+    raise ActionController::RoutingError, "invalid account type for new: #{account_type}"
+  end
+
+  # Expects `options[:account]`; returns account_type class name string if
+  # options[:account] object is a valid account_type; otherwise raises an
+  # AccountController::RoutingError exception.
+  def valid_account_type_for_existing!(account)
+    account_type = account.class.to_s
+    return account_type if Account.config.account_types.include?(account_type)
+    raise ActionController::RoutingError, "invalid account type for existing: #{account_type}"
+  end
+
+  # Build a new subclassed `Account` object .
+  # Assumes `account_type` is a valid class name; and assumes params are
+  # whitelisted/safe to assign.
+  def new_account(account_type)
+    account_type.constantize.new
+  end
+
+  # Assumes params are safe. Uses mass assignment (which is dangerous). To
+  # ensure safety we should either pass strong_params to the builder or
+  # whitelist params within the builder via the `parse_params_...` methods.
+  def assign_params(account, params, params_key)
+    account.assign_attributes(params[params_key])
+    account
+  end
+
+  # Build an owner account_user.
+  # Returns the modified account object.
+  def build_account_users(account, current_user, owner_user)
+    account.account_users.build(
+      user_id: owner_user.id,
+      user_role: 'Owner',
+      created_by: current_user.id,
+    )
+    account
+  end
+
+  # Set the affiliate for an account only if affiliates are supported for the
+  # given account type.  Returns the modified account object.
+  # Silently fails if affiliate id doesn't exist; add validations to the
+  # subclassed `Account` if you'd like to require an affiliate.
+  def set_affiliate(account, params, params_key)
+    return account unless account.class.using_affiliate?
+    id = params.fetch(params_key, {}).fetch(:affiliate_id, nil)
```

Comment:
Seeing this again, it seems like if it's possible, the `@params` should be just the `params[account_type]`. Passing around the `params_key` is seeming very confusing and possibly doing redundant lookups.